### PR TITLE
Refactor source file structure

### LINKED ISSUE #977: Proposal: Refactor source file structure
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

As mentioned in https://github.com/apache/incubator-kvrocks/pull/973#pullrequestreview-1137848493, we can categorize source files in to different folders to modularize the source structure.

### Solution

The preliminary plan is as follows:

```
common/
  event_util fd_util parse_util cron encoding db_util event_util log_collector
  parse_util rw_lock status tls_util util

database/
  batch_extractor compact_filter compaction_checker event_listener geohash
  lock_manager redis_db redis_disk redis_metadata redis_pubsub
  scripting stats storage table_properties_collector

service/
  server worker redis_reply redis_connection redis_request

distributed/
  cluster redis_slot replication slot_import slot_migrate 

data_types/
  redis_bitmap redis_bitmap_string redis_geo redis_list redis_hash redis_set
  redis_sortedint redis_stream_base redis_stream redis_string redis_zset

commands/
  redis_cmd (need to be divided)

config/
  config config_type config_util
  
crypto/
  rand sha1 rocksdb_crc32c

```

Since I am not very familiar with certain parts of the code, I welcome everyone's thoughts, comments and suggestions!

cc @git-hulk @ShooterIT @caipengbo @tisonkun @torwig 

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by PragmaTwice**
Mostly follows https://github.com/apache/incubator-kvrocks/issues/977#issuecomment-1278446623, except:

- `log_collector` and `tls_util` is moved to `network/`
- add `task_runner` to `common/`

### CODE REVIEW

**Comment by caipengbo** [comment]
`log_collector` is not a network thing, and i think it should be put under `common`

**Comment by PragmaTwice** [comment]
> `log_collector` is not a network thing, and i think it should be put under `common`

I think both `common` and `network` is ok, I put it into `network` because I find that this file relies on `redis_reply` and is using redis protocol. I will move it to `common`.

**Comment by caipengbo** [comment]
@PragmaTwice  log_collector is not a network-related thing, it is a statistical thing, how about adding a `stats` directory with `log_collector.h` and `stats.h` or even `redis_disk.h` (or can rename it `disk_stats.h`?)

**Comment by PragmaTwice** [comment]
> @PragmaTwice log_collector is not a network-related thing, it is a statistical thing, how about adding a `stats` directory with `log_collector.h` and `stats.h` or even `redis_disk.h` (or can rename it `disk_stats.h`?)

Greate idea. I will do it.

**Comment by caipengbo** [comment]
`geohash.h` file appears to be used only by `redis_geo.h`. I don't think it's appropriate to put it in `storage`. Can we just put it inside `types`?

**Comment by PragmaTwice** [comment]
Done. Current structure:

```
$ tree src
├── cluster
│   ├── cluster.cc
│   ├── cluster.h
│   ├── redis_slot.cc
│   ├── redis_slot.h
│   ├── replication.cc
│   ├── replication.h
│   ├── slot_import.cc
│   ├── slot_import.h
│   ├── slot_migrate.cc
│   └── slot_migrate.h
├── commands
│   ├── redis_cmd.cc
│   └── redis_cmd.h
├── common
│   ├── cron.cc
│   ├── cron.h
│   ├── db_util.h
│   ├── encoding.cc
│   ├── encoding.h
│   ├── event_util.h
│   ├── fd_util.h
│   ├── parse_util.h
│   ├── rand.cc
│   ├── rand.h
│   ├── rocksdb_crc32c.h
│   ├── rw_lock.h
│   ├── scope_exit.h
│   ├── sha1.cc
│   ├── sha1.h
│   ├── status.h
│   ├── task_runner.cc
│   ├── task_runner.h
│   ├── util.cc
│   └── util.h
├── config
│   ├── config.cc
│   ├── config.h
│   ├── config_type.h
│   ├── config_util.cc
│   └── config_util.h
├── main.cc
├── network
│   ├── redis_connection.cc
│   ├── redis_connection.h
│   ├── redis_reply.cc
│   ├── redis_reply.h
│   ├── redis_request.cc
│   ├── redis_request.h
│   ├── server.cc
│   ├── server.h
│   ├── tls_util.cc
│   ├── tls_util.h
│   ├── worker.cc
│   └── worker.h
├── stats
│   ├── disk_stats.cc
│   ├── disk_stats.h
│   ├── log_collector.cc
│   ├── log_collector.h
│   ├── stats.cc
│   └── stats.h
├── storage
│   ├── batch_extractor.cc
│   ├── batch_extractor.h
│   ├── compact_filter.cc
│   ├── compact_filter.h
│   ├── compaction_checker.cc
│   ├── compaction_checker.h
│   ├── event_listener.cc
│   ├── event_listener.h
│   ├── geohash.cc
│   ├── geohash.h
│   ├── lock_manager.cc
│   ├── lock_manager.h
│   ├── redis_db.cc
│   ├── redis_db.h
│   ├── redis_metadata.cc
│   ├── redis_metadata.h
│   ├── redis_pubsub.cc
│   ├── redis_pubsub.h
│   ├── scripting.cc
│   ├── scripting.h
│   ├── storage.cc
│   ├── storage.h
│   ├── table_properties_collector.cc
│   └── table_properties_collector.h
├── types
│   ├── redis_bitmap.cc
│   ├── redis_bitmap.h
│   ├── redis_bitmap_string.cc
│   ├── redis_bitmap_string.h
│   ├── redis_geo.cc
│   ├── redis_geo.h
│   ├── redis_hash.cc
│   ├── redis_hash.h
│   ├── redis_list.cc
│   ├── redis_list.h
│   ├── redis_set.cc
│   ├── redis_set.h
│   ├── redis_sortedint.cc
│   ├── redis_sortedint.h
│   ├── redis_stream_base.cc
│   ├── redis_stream_base.h
│   ├── redis_stream.cc
│   ├── redis_stream.h
│   ├── redis_string.cc
│   ├── redis_string.h
│   ├── redis_zset.cc
│   └── redis_zset.h
├── valgrind.sup
└── version.h.in

8 directories, 104 files
```

**Comment by PragmaTwice** [comment]
> `geohash.h` file appears to be used only by `redis_geo.h`. I don't think it's appropriate to put it in `storage`. Can we just put it inside `types`?

Agree with that.

**Comment by caipengbo** [comment]
> log_collector is not a network-related thing, it is a statistical thing, how about adding a stats directory with log_collector.h and stats.h or even redis_disk.h (or can rename it disk_stats.h?)

We renamed `redis_disk.h` to `disk_stats.h`, do you think that is reasonable? @tanruixiang

**Comment by PragmaTwice** [comment]
Current source structure:

```
$ tree src
├── cluster
│   ├── cluster.cc
│   ├── cluster.h
│   ├── redis_slot.cc
│   ├── redis_slot.h
│   ├── replication.cc
│   ├── replication.h
│   ├── slot_import.cc
│   ├── slot_import.h
│   ├── slot_migrate.cc
│   └── slot_migrate.h
├── commands
│   ├── redis_cmd.cc
│   └── redis_cmd.h
├── common
│   ├── cron.cc
│   ├── cron.h
│   ├── db_util.h
│   ├── encoding.cc
│   ├── encoding.h
│   ├── event_util.h
│   ├── fd_util.h
│   ├── parse_util.h
│   ├── rand.cc
│   ├── rand.h
│   ├── rocksdb_crc32c.h
│   ├── rw_lock.h
│   ├── scope_exit.h
│   ├── sha1.cc
│   ├── sha1.h
│   ├── status.h
│   ├── task_runner.cc
│   ├── task_runner.h
│   ├── util.cc
│   └── util.h
├── config
│   ├── config.cc
│   ├── config.h
│   ├── config_type.h
│   ├── config_util.cc
│   └── config_util.h
├── main.cc
├── network
│   ├── redis_connection.cc
│   ├── redis_connection.h
│   ├── redis_reply.cc
│   ├── redis_reply.h
│   ├── redis_request.cc
│   ├── redis_request.h
│   ├── server.cc
│   ├── server.h
│   ├── tls_util.cc
│   ├── tls_util.h
│   ├── worker.cc
│   └── worker.h
├── stats
│   ├── disk_stats.cc
│   ├── disk_stats.h
│   ├── log_collector.cc
│   ├── log_collector.h
│   ├── stats.cc
│   └── stats.h
├── storage
│   ├── batch_extractor.cc
│   ├── batch_extractor.h
│   ├── compact_filter.cc
│   ├── compact_filter.h
│   ├── compaction_checker.cc
│   ├── compaction_checker.h
│   ├── event_listener.cc
│   ├── event_listener.h
│   ├── lock_manager.cc
│   ├── lock_manager.h
│   ├── redis_db.cc
│   ├── redis_db.h
│   ├── redis_metadata.cc
│   ├── redis_metadata.h
│   ├── redis_pubsub.cc
│   ├── redis_pubsub.h
│   ├── scripting.cc
│   ├── scripting.h
│   ├── storage.cc
│   ├── storage.h
│   ├── table_properties_collector.cc
│   └── table_properties_collector.h
├── types
│   ├── geohash.cc
│   ├── geohash.h
│   ├── redis_bitmap.cc
│   ├── redis_bitmap.h
│   ├── redis_bitmap_string.cc
│   ├── redis_bitmap_string.h
│   ├── redis_geo.cc
│   ├── redis_geo.h
│   ├── redis_hash.cc
│   ├── redis_hash.h
│   ├── redis_list.cc
│   ├── redis_list.h
│   ├── redis_set.cc
│   ├── redis_set.h
│   ├── redis_sortedint.cc
│   ├── redis_sortedint.h
│   ├── redis_stream_base.cc
│   ├── redis_stream_base.h
│   ├── redis_stream.cc
│   ├── redis_stream.h
│   ├── redis_string.cc
│   ├── redis_string.h
│   ├── redis_zset.cc
│   └── redis_zset.h
├── valgrind.sup
└── version.h.in

8 directories, 104 files
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -148,12 +148,12 @@ find_library(FOUND_UNWIND_LIB unwind)
 set(WARNING_FLAGS -Wall -Wpedantic -Wsign-compare -Wreturn-type)
 
 # kvrocks objects target
-file(GLOB KVROCKS_SRCS src/*.cc)
+file(GLOB_RECURSE KVROCKS_SRCS src/*.cc)
```
Comment (tisonkun) [inline CMakeLists.txt:151]: Does the build benefit from separating compile units? For example, one `.o` per directory and for better incremental compilation. If it does we can implement it here or as a follow-up.

File: CMakeLists.txt [hunk 2]
```diff
@@ -148,12 +148,12 @@ find_library(FOUND_UNWIND_LIB unwind)
...
 list(FILTER KVROCKS_SRCS EXCLUDE REGEX src/main.cc)
 
 add_library(kvrocks_objs OBJECT ${KVROCKS_SRCS})
 
-target_include_directories(kvrocks_objs PUBLIC src ${PROJECT_BINARY_DIR})
+target_include_directories(kvrocks_objs PUBLIC src src/common ${PROJECT_BINARY_DIR})
```
Comment (tisonkun) [inline CMakeLists.txt:156]: What's the reason of this change?

File: src/cluster/cluster.cc [hunk 3]
```diff
@@ -25,10 +25,10 @@
 #include <cstring>
 #include <memory>
 
+#include "commands/redis_cmd.h"
```
Comment (tisonkun) [inline src/cluster/cluster.cc:28]: Does C++ module style prefer UpperCase? That is, "Command/redis_cmd.h".

Comment (tisonkun) [reply src/cluster/cluster.cc:28]: Besides, prefer "command" to "commands".

**Review by tisonkun — COMMENTED** [review]
Great work! Comments inline.

File: src/cluster/cluster.cc (hunk 3 above)
Comment (PragmaTwice) [reply src/cluster/cluster.cc:28]: I think we need to unify the style of filenames, if it is `Command`, maybe it will be `Command/RedisCmd.h`, so I think we can keep underscore-style.

Comment (tisonkun) [reply src/cluster/cluster.cc:28]: > I think we need to unify the style of filenames, if it is Command, maybe it will be Command/RedisCmd.h, so I think we can keep underscore-style.

Make sense :)

File: CMakeLists.txt (hunk 2 above)
Comment (PragmaTwice) [reply CMakeLists.txt:156]: Since header files in `common` is frequently used, so e.g. it can be just `"status.h"` rather than `"common/status.h"`.

File: CMakeLists.txt (hunk 1 above)
Comment (PragmaTwice) [reply CMakeLists.txt:151]: I think there is not much dependency between translation units. In theory, if there are 1000 source files, a 1000-core cpu and enough memory, we can compile completely in parallel.

File: src/cluster/cluster.cc (hunk 3 above)
Comment (git-hulk) [reply src/cluster/cluster.cc:28]: > Does C++ module style prefer UpperCase? That is, "Command/redis_cmd.h".

ClickHouse src dir uses uppercase but some of dir is the snake case, I can't say which one is better, but I prefer the snake case than the uppercase from my own habits.

**Comment by tanruixiang** [comment]
> We renamed `redis_disk.h` to `disk_stats.h`, do you think that is reasonable? @tanruixiang

Yes, I think it's reasonable.

**Comment by git-hulk** [comment]
@PragmaTwice After looking detail, it seems `server` is better than `network` to replace `service`. It's my bad that didn't mention this option, others are good to me. I think we can merge it as soon as possible to avoid conflicts.

**Comment by PragmaTwice** [comment]
> @PragmaTwice After looking detail, it seems `server` is better than `network` to replace `service`. It's my bad that didn't mention this option, others are good to me. I think we can merge it as soon as possible to avoid conflicts.

Done. @git-hulk 

The current diff in github review page seems to be a mess, maybe because I rebase some commits.

**Comment by PragmaTwice** [comment]
> The current diff in github review page seems to be a mess, maybe because I rebase some commits.

Solved by re-rebase and force-push.

Current source structure:
```
src
├── cluster
│   ├── cluster.cc
│   ├── cluster.h
│   ├── redis_slot.cc
│   ├── redis_slot.h
│   ├── replication.cc
│   ├── replication.h
│   ├── slot_import.cc
│   ├── slot_import.h
│   ├── slot_migrate.cc
│   └── slot_migrate.h
├── commands
│   ├── redis_cmd.cc
│   └── redis_cmd.h
├── common
│   ├── cron.cc
│   ├── cron.h
│   ├── db_util.h
│   ├── encoding.cc
│   ├── encoding.h
│   ├── event_util.h
│   ├── fd_util.h
│   ├── parse_util.h
│   ├── rand.cc
│   ├── rand.h
│   ├── rocksdb_crc32c.h
│   ├── rw_lock.h
│   ├── scope_exit.h
│   ├── sha1.cc
│   ├── sha1.h
│   ├── status.h
│   ├── task_runner.cc
│   ├── task_runner.h
│   ├── util.cc
│   └── util.h
├── config
│   ├── config.cc
│   ├── config.h
│   ├── config_type.h
│   ├── config_util.cc
│   └── config_util.h
├── main.cc
├── server
│   ├── redis_connection.cc
│   ├── redis_connection.h
│   ├── redis_reply.cc
│   ├── redis_reply.h
│   ├── redis_request.cc
│   ├── redis_request.h
│   ├── server.cc
│   ├── server.h
│   ├── tls_util.cc
│   ├── tls_util.h
│   ├── worker.cc
│   └── worker.h
├── stats
│   ├── disk_stats.cc
│   ├── disk_stats.h
│   ├── log_collector.cc
│   ├── log_collector.h
│   ├── stats.cc
│   └── stats.h
├── storage
│   ├── batch_extractor.cc
│   ├── batch_extractor.h
│   ├── compact_filter.cc
│   ├── compact_filter.h
│   ├── compaction_checker.cc
│   ├── compaction_checker.h
│   ├── event_listener.cc
│   ├── event_listener.h
│   ├── lock_manager.cc
│   ├── lock_manager.h
│   ├── redis_db.cc
│   ├── redis_db.h
│   ├── redis_metadata.cc
│   ├── redis_metadata.h
│   ├── redis_pubsub.cc
│   ├── redis_pubsub.h
│   ├── scripting.cc
│   ├── scripting.h
│   ├── storage.cc
│   ├── storage.h
│   ├── table_properties_collector.cc
│   └── table_properties_collector.h
├── types
│   ├── geohash.cc
│   ├── geohash.h
│   ├── redis_bitmap.cc
│   ├── redis_bitmap.h
│   ├── redis_bitmap_string.cc
│   ├── redis_bitmap_string.h
│   ├── redis_geo.cc
│   ├── redis_geo.h
│   ├── redis_hash.cc
│   ├── redis_hash.h
│   ├── redis_list.cc
│   ├── redis_list.h
│   ├── redis_set.cc
│   ├── redis_set.h
│   ├── redis_sortedint.cc
│   ├── redis_sortedint.h
│   ├── redis_stream_base.cc
│   ├── redis_stream_base.h
│   ├── redis_stream.cc
│   ├── redis_stream.h
│   ├── redis_string.cc
│   ├── redis_string.h
│   ├── redis_zset.cc
│   └── redis_zset.h
├── valgrind.sup
└── version.h.in

8 directories, 104 files
```

**Review by git-hulk — DISMISSED** [review]

File: src/commands/redis_cmd.cc [hunk 4]
```diff
@@ -32,30 +32,30 @@
 #include <unordered_map>
```
Comment (tisonkun) [inline src/commands/redis_cmd.cc:32]: I'd recall naming this folder as `command`. What do you think?

Comment (torwig) [reply src/commands/redis_cmd.cc:32]: @tisonkun  Do you mean singular `command` instead of plural `commands`? If yes, then we have `types`, `tools` and `tests`, so `commands` is OK in my opinion.

Comment (tisonkun) [reply src/commands/redis_cmd.cc:32]: @torwig well. Since Kvrocks code is almost internal, that users talk to the server via wire protocol, I'm OK with either choice.

**Review by tisonkun — DISMISSED** [review]

**Review by torwig — DISMISSED** [review]
@PragmaTwice Good job!
LGTM (with or without `src/common` among include directories)

**Comment by tisonkun** [comment]
@PragmaTwice please rebase on the latest unstable branch.

@ShooterIT @caipengbo could you give a review on this patch? Perhaps we should move it forward in a timely manner to avoid further conflicts.

**Review by caipengbo — DISMISSED** [review]
LGTM

**Comment by PragmaTwice** [comment]
> @PragmaTwice please rebase on the latest unstable branch.

Done.

**Review by git-hulk — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Review by tisonkun — APPROVED** [review]
